ext/cl_khr_defined_builtin_kernels.asciidoc

```diff
@@ -99,222 +103,414 @@ accelerators while being able to synchronize and optimize data transfers at
-4+| OpenCL C Semantics
-4+a|
-[source,c]
-----
-__kernel void __khr_io_stream_out_uchar(
-   uchar *input, size_t *num) {
-   // It is not feasible to describe this kernel in OpenCL C as I/O devices
-   // are not representable with it.
-}
-----
+Fields of the `cl_dbk_leaky_relu_v1` structure:
+. cl_tensor_desc_t in: Input tensor description.
+. cl_tensor_desc_t out: Output tensor description.
+. cl_float alpha: Coefficient of leakage.
+| *Kernel arguments*
+a|
+. cl_tensor_t in: The input tensor.
+. cl_tensor_t out: The output tensor.
+| *Description*
+a|
+Applies operation `alpha * x if x < 0 else x` on all
+elements of the `in` tensor.
+
+If target device does not support denormals, then `alpha` is flushed
+to zero before the operation is applied.
 
-| *khr_io_stream_in_blocking_uchar*
-| Blocking read of data from a sensor/stream associated with the device.
-a| -
+| *Kernel mode properties*
+| N/A
+| *Attribute validation rules*
 a|
-[start=0]
-. uchar* [out]: The data.
-* size_t* [in]: How many bytes to read before returning.
+* `shapeof(in) == shapeof(out)`
+* `elementof(in) == elementof(out)`
+* `alpha` must be a finite value.
+|===
+
+==== Launching DBKs from the Device Side ====
+
+DBKs are primarily meant to be launched as kernel commands via
+host-side command queues.  Optionally, they can be callable from
+device-side via `enqueue_kernel`:
+
+TBC. This probably needs device-side function corresponding to
+clCreateDefinedBuiltInKernelDescriptor.
+
+==== Sample Code ====
 
-4+| OpenCL C Semantics
-4+a|
 [source,c]
 ----
-__kernel void __khr_io_stream_in_blocking_uchar(uchar *output, size_t *num) {
-   while (*num) {
-       size_t num_read = *num;
-       __khr_io_stream_in_uchar(output, &num_read);
-       num -= num_read;
-       output += num_read;
-   }
+// TBD. Similarly in cl_qcom_ml_ops, tensors have type
+// (cl_channel_type) and a number of dimensions (rank) and dimension
+// sizes (shape). Difference over the cl_qcom_ml_ops is that the rank is
+// "unlimited".
+cl_tensor_desc_t lhs_tensor_desc = TBD;
+cl_tensor_desc_t rhs_tensor_desc = TBD;
+cl_tensor_desc_t res_tensor_desc = TBD;
+
+cl_dkb_attributes_matmul_v1 matmul_attrs = {
+  lhs_tensor_desc, rhs_tensor_desc, res_tensor_desc,
+  1, 0 // = Transpose lhs tensor
 }
-----
 
-|===
-
-==== Launching BiKs from the Device Side ====
-
-BiKs are primarily meant to be launched as kernel commands via host-side command queues.
-Optionally, they can be callable from device-side via
-`enqueue_kernel`: This capability can be queried on per BiK basis at compile-time in OpenCL C by checking for macro definitions which has the following naming convention: `cl_khr_bik_BUILTIN_KERNEL_NAME`. In case a BiK macro is defined, a kernel with a naming convention `__khr_BUILTIN_KERNEL_NAME()` can be enqueued by the program at device side as software-defined kernels.
+cl_dbk_mode_properties matmul_props = {
+  // Request a matmul implementation that meets this precision.
```

Comment:
Thinking of this, we have two use cases: 1) Request a list of DBKs from the device with at least a given precision. 2) Enqueuing a DBK with reduced precision requirements (which can help DBK implementation to go faster). If we combine these two, the handle then includes the reduced precision for all invocations of the DBK with that handle. Should be fine for most use cases, and we can request multiple handles.